models/gif_pop_psc_exp.cpp

```diff
@@ -0,0 +1,702 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/* Point process population model with exponential postsynaptic currents and
+ * adaptation */
+
+/* [1]: line numbers in comments refer to the algorithm pseudocode in 
+Figures 11 and 12 of the paper
+Schwalger T, Deger M, Gerstner W (2017) Towards a theory of cortical columns: 
+From spiking neurons to interacting neural populations of finite size. 
+PLoS Comput Biol 13(4): e1005507. 
+https://doi.org/10.1371/journal.pcbi.1005507 */
+
+#include "gif_pop_psc_exp.h"
+#include "universal_data_logger_impl.h"
+#include "compose.hpp"
+
+namespace nest
+{
+/* ----------------------------------------------------------------
+ * Recordables map
+ * ---------------------------------------------------------------- */
+
+RecordablesMap< gif_pop_psc_exp > gif_pop_psc_exp::recordablesMap_;
+
+// Override the create() method with one call to RecordablesMap::insert_()
+// for each quantity to be recorded.
+template <>
+void
+RecordablesMap< gif_pop_psc_exp >::create()
+{
+  // use standard names whereever you can for consistency!
+  insert_( names::V_m, &gif_pop_psc_exp::get_V_m_ );
+  insert_( names::n_events, &gif_pop_psc_exp::get_n_events_ );
+  insert_( names::E_sfa, &gif_pop_psc_exp::get_E_sfa_ );
+  insert_( names::mean, &gif_pop_psc_exp::get_mean_ );
+  insert_( names::I_syn_ex, &gif_pop_psc_exp::get_I_syn_ex_ );
+  insert_( names::I_syn_in, &gif_pop_psc_exp::get_I_syn_in_ );
+}
+
+/* ----------------------------------------------------------------
+ * Default constructors defining default parameters and state
+ * ---------------------------------------------------------------- */
+
+nest::gif_pop_psc_exp::Parameters_::Parameters_()
+  : N_( 100 )          // 1
+  , tau_m_( 20. )      // ms
+  , c_m_( 250. )       // pF
+  , t_ref_( 4. )       // ms
+  , lambda_0_( 10. )   // 1/s
+  , Delta_V_( 2. )     // mV
+  , len_kernel_( -1 )  // time steps
+  , I_e_( 0. )         // pA
+  , V_reset_( 0. )     // mV
+  , V_T_star_( 15. )   // mV
+  , E_L_( 0. )         // mV
+  , tau_syn_ex_( 3.0 ) // ms
+  , tau_syn_in_( 6.0 ) // ms
+  , BinoRand_( true )  // bool
+{
+  tau_sfa_.clear();
+  q_sfa_.clear();
+  tau_sfa_.push_back( 300.0 ); // ms
+  q_sfa_.push_back( 0.5 );     // mV
+}
+
+nest::gif_pop_psc_exp::State_::State_()
+  : y0_( 0.0 )
+  , I_syn_ex_( 0.0 )
+  , I_syn_in_( 0.0 )
+  , V_m_( 0.0 )
+  , theta_hat_( 0.0 )   // initialization value has no effect for theta_hat_
```

Comment:
`theta_hat_` should be initialized after `n_spikes_` to avoid warnings during compilation.